src/mlpack/methods/ann/layer/sigmoid_cross_entropy_error_impl.hpp

```diff
@@ -0,0 +1,76 @@
+/**
+ * @file sigmoid_cross_entropy_error_impl.hpp
+ * @author Kris Singh
+ *
+ * Implementation of the sigmoid cross entropy error performance function.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_SIGMOID_CROSS_ENTROPY_ERROR
+#define MLPACK_METHODS_ANN_LAYER_SIGMOID_CROSS_ENTROPY_ERROR
+
+// In case it hasn't yet been included.
+#include "sigmoid_cross_entropy_error.hpp"
+#include <mlpack/methods/ann/activation_functions/softplus_function.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+SigmoidCrossEntropyError<InputDataType, OutputDataType>
+::SigmoidCrossEntropyError()
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+inline double SigmoidCrossEntropyError<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>&& input, const arma::Mat<eT>&& target)
+{
+  arma::uvec positive = arma::find(input > 0);
+  arma::mat output;
+  SoftplusFunction::Fn(static_cast<arma::mat>(-arma::abs(input)), output);
+  double loss = arma::accu(input(positive)) - arma::accu(input % target) +
+        arma::accu(output);
```

Comment:
I think you could optimize this to avoid the `positive` vector, maybe like this:

```
double loss = arma::accu(arma::max(input, 0)) - arma::accu(input % target) + arma::accu(output);
```

I am not 100% sure that `arma::accu(arma::max(input, 0))` will avoid creating a temporary, but if it doesn't, you could also just loop over the `input` matrix.

Reply:
Could you explain arma::accu(arma::max(input, 0)). This  would return a max number along every column whereas arma::find(input > 0) find the indicies of every non-negative input. So, how could one replace another.

Reply:
Ah yeah, sorry, I used the wrong function here.  Instead of `arma::accu(arma::max(input, 0))` I meant `arma::accu(arma::clamp(input, 0, std::numeric_limits<double>::max()))`.

Reply:
@rcurtin `arma::clamp` does create a copy of the original matrix. How do you wish to proceed here?

Reply:
@ShikharJ: are you sure that the combination of `accu(clamp(...))` will create a temporary matrix?  It looks like `clamp()` returns an `arma::mtOp<...>` which signifies that the clamp operation should be done, but I think then `accu()` will use an `arma::Proxy<>` object, which I think in turn will create a temporary for an `mtOp<>`.  The best way to check for sure would be to create a simple program, compile with `-DARMA_EXTRA_DEBUG`, and parse through the output to see if the constructor for `arma::Mat<eT>` is called during the `accu(clamp())` operation.

In any case, if you do confirm that this makes a copy, you can just write a loop over the input matrix and manually clamp the elements to be greater than zero as you sum them.

Reply:
The `Armadillo` documentation mentions this about  `clamp()`:
```
Create a copy of X with each element clamped to be between min_val and max_val
```
`accu()` should just return the sum of the intended elements. So all in all, `accu(clamp(...))` will internally use a temporary, but just return the sum in the end. Am I misunderstanding the context of your concern?

Reply:
Yes, so the documentation is not misleading but it doesn't mention that Armadillo is sometimes able to take advantage of expressions and avoid copies.  Based on what I saw in the Armadillo code, Armadillo is not currently able to avoid a copy in this case, so if it is ok with you, I think it would be better to manually write a loop to calculate the sum here.

Reply:
Ah, I see in #1200 that you already wrote the loop.  Thanks! :)